### PR TITLE
Save/Load config as yaml

### LINKED ISSUE #4: [CODE IMPROVEMENT] Save cfg as yaml instead of dill/pickle
### 🔧 Proposed code refactoring

Move from pickle format of saving `cfg_last.p` to yaml (or similar format).

### Motivation

- Human readable data format
- Better comparability to possible code changes (pickle breaks easily)
- No security concerns when sharing configurations
- Saving/Ingestion to `config.yaml` is already implemented in the UI

### Note

`cfg_last.p` stores

```
['_seed',
'_stop_words_ids',
 '_tokenizer_sep_token',
 '_tokenizer_mask_token_id',
 '_tokenizer_eos_token']
```
that is currently not stored in `config.yaml`. Seed is stored in the logs, while the remaining attributes are created in `get_tokenizer` method.

### PR DESCRIPTION
**Pull request by maxjeblick**
This PR converts pickled configurations into yaml configuration.

It introduces a breaking change to the current UI (meaning that old experiments won't work). As the repo is in its initial phase, I did not add any conversion scripts to load pickle files and convert them into yamls to keep the repo clean.

Fixes  #4

### CODE REVIEW

**Comment by psinger** [comment]
Add also an option to pass a yaml via CLI

File: train_wave.py [hunk 1]
```diff
@@ -89,10 +86,9 @@ def check_for_done(process_queue):
     from train import run
 
     if "config" in parser_args:
-        cfg = load_config(parser_args.config)
-    elif "pickle" in parser_args:
-        with open(parser_args.pickle, "rb") as pickle_file:
-            cfg = dill.load(pickle_file)
+        cfg = load_config_py(parser_args.config)
```
Comment (pascal-pfeiffer) [inline train_wave.py:89]: mark as deprecated? or should we keep it?
Actually, it is never used here, so should be good for removal even.

File: train.py [hunk 2]
```diff
@@ -632,16 +631,13 @@ def run(cfg: Any) -> None:
...
-    )
+    parser.add_argument("-Y", "--yaml", help="yaml filename", default=argparse.SUPPRESS)
     parser_args, unknown = parser.parse_known_args(sys.argv)
 
     if "config" in parser_args:
-        cfg = load_config(parser_args.config)
-    elif "pickle" in parser_args:
-        with open(parser_args.pickle, "rb") as pickle_file:
-            cfg = dill.load(pickle_file)
+        cfg = load_config_py(parser_args.config)
```
Comment (pascal-pfeiffer) [inline train.py:638]: mark as deprecated? or should we keep it?

**Review by pascal-pfeiffer — CHANGES_REQUESTED** [review]
Thank you @maxjeblick 
Big changes that will make sharing of configs easier. 

I am running into a few issues still when updating an existing instance

opening the chat
```
Traceback (most recent call last):
  File "/home/pascal/h2o-llmstudio/./app_utils/handlers.py", line 308, in handle
    await experiment_display(q)
  File "/home/pascal/h2o-llmstudio/./app_utils/sections/experiment.py", line 1065, in experiment_display
    cfg, model, tokenizer = load_cfg_model_tokenizer(
  File "/home/pascal/h2o-llmstudio/./app_utils/sections/experiment.py", line 1596, in load_cfg_model_tokenizer
    tokenizer = get_tokenizer(cfg)
  File "/home/pascal/h2o-llmstudio/./llm_studio/src/datasets/text_utils.py", line 69, in get_tokenizer
    filter(None, cfg.prediction.stop_tokens.split(","))
AttributeError: 'list' object has no attribute 'split'
```

Starting from an old experiment:
```
Traceback (most recent call last):
  File "/home/pascal/h2o-llmstudio/./app_utils/handlers.py", line 178, in handle
    await experiment_start(q)
  File "/home/pascal/h2o-llmstudio/./app_utils/sections/experiment.py", line 443, in experiment_start
    option_items = get_ui_elements(cfg=q.client["experiment/start/cfg"], q=q)
  File "/home/pascal/h2o-llmstudio/./app_utils/utils.py", line 1069, in get_ui_elements
    elements_group = get_ui_elements(cfg=v, q=q, limit=limit, pre=pre)
  File "/home/pascal/h2o-llmstudio/./app_utils/utils.py", line 1053, in get_ui_elements
    t = get_ui_element(
  File "/home/pascal/h2o-llmstudio/./app_utils/utils.py", line 740, in get_ui_element
    ui.textbox(
  File "/home/pascal/.local/share/virtualenvs/h2o-llmstudio-tT3gHl3a/lib/python3.10/site-packages/h2o_wave/ui.py", line 437, in textbox
    return Component(textbox=Textbox(
  File "/home/pascal/.local/share/virtualenvs/h2o-llmstudio-tT3gHl3a/lib/python3.10/site-packages/h2o_wave/types.py", line 1015, in __init__
    _guard_scalar('Textbox.value', value, (str,), False, True, False)
  File "/home/pascal/.local/share/virtualenvs/h2o-llmstudio-tT3gHl3a/lib/python3.10/site-packages/h2o_wave/types.py", line 37, in _guard_scalar
    raise ValueError(f'{name}: want one of {types}, got {type(value)}')
ValueError: Textbox.value: want one of (<class 'str'>,), got <class 'list'>
```

When running the CLI pipeline with an exported yaml:
```
Traceback (most recent call last):
  File "/home/pascal/h2o-llmstudio/train.py", line 640, in <module>
    cfg = load_config_yaml(parser_args.yaml)
  File "/home/pascal/h2o-llmstudio/llm_studio/src/utils/config_utils.py", line 210, in load_config_yaml
    output_directory=cfg_dict["output_directory"],
KeyError: 'output_directory'
```

I think some may be related to other changes, like the recent tokenizer change. 

Also, I noticed that the Dataset was removed when switching between commits. 

Overall, our migration strategy will need to be taken either more carefully, or we should ignore completely and mention in the README that the app database can crash with every new commit and pulling new versions should be done with care. 

Please also update the README regarding usage of the CLI.

**Comment by psinger** [comment]
Shall we maybe also add a NEW section to the Readme updating about more major changes?

**Comment by maxjeblick** [comment]
> Shall we maybe also add a NEW section to the Readme updating about more major changes?

Yes changelog makes sense.

File: train.py (hunk 2 above)
Comment (maxjeblick) [reply train.py:641]: We can remove it once the kaggle/colab kernels have been updated

**Comment by maxjeblick** [comment]
@pascal-pfeiffer could you share the configuration files you had issues with?

File: train.py (hunk 2 above)
Comment (pascal-pfeiffer) [reply train.py:641]: Sounds good to have the option. Let's keep it then

**Comment by pascal-pfeiffer** [comment]
> @pascal-pfeiffer could you share the configuration files you had issues with?
Did we save stop_tokens as a list at some point?

I think, I had some mix in my install. Seems to work just fine with the latest commit. Let me do one more review and I think we are good to merge it.

File: llm_studio/src/utils/config_utils.py [hunk 3]
```diff
@@ -61,3 +78,152 @@ def load_config(config_path: str, config_name: str = "Config"):
...
+        else:
+            raise _get_type_annotation_error(v, type_annotations[k])
+
+        items += t
+
+    return items
+
+
+def save_config_yaml(path: str, cfg: ConfigProblemBase) -> None:
+    """Saves config as dill file
```
Comment (pascal-pfeiffer) [inline llm_studio/src/utils/config_utils.py:185]: yaml

File: llm_studio/src/utils/config_utils.py [hunk 4]
```diff
@@ -61,3 +78,152 @@ def load_config(config_path: str, config_name: str = "Config"):
...
+        ),
+        training=ConfigNLPCausalLMTraining.from_dict(cfg_dict.get("training", {})),
+        prediction=ConfigNLPCausalLMPrediction.from_dict(
+            cfg_dict.get("prediction", {})
+        ),
+        environment=ConfigNLPCausalLMEnvironment.from_dict(
+            cfg_dict.get("environment", {})
+        ),
+        logging=ConfigNLPCausalLMLogging.from_dict(cfg_dict.get("logging", {})),
+    )
```
Comment (pascal-pfeiffer) [inline llm_studio/src/utils/config_utils.py:228]: any way to have that dynamically grab the problem type? We will need to consider if we add other configs. Likely need to add to the yaml, too.

**Review by pascal-pfeiffer — CHANGES_REQUESTED** [review]
Thank you again @maxjeblick 

I only found one more small issue, when running on a fresh environment after running

```sh
make clean-data
make clean-output
make clean-env
make setup
```

Error when starting an experiment:
```
Traceback (most recent call last):

  File "/home/pascal/h2o-llmstudio/./app_utils/handlers.py", line 163, in handle
    await experiment_run(q, pre="experiment/start")

  File "/home/pascal/h2o-llmstudio/./app_utils/sections/experiment.py", line 503, in experiment_run
    start_experiment(cfg=cfg, q=q, pre=pre)

  File "/home/pascal/h2o-llmstudio/./app_utils/utils.py", line 1549, in start_experiment
    cfg = copy_config(cfg)

  File "/home/pascal/h2o-llmstudio/./app_utils/utils.py", line 1854, in copy_config
    save_config_yaml(tmp_file, cfg)

  File "/home/pascal/h2o-llmstudio/./llm_studio/src/utils/config_utils.py", line 195, in save_config_yaml
    with open(path, "w") as fp:

FileNotFoundError: [Errno 2] No such file or directory: 'output/fd7e37fa-c593-44f5-b62d-2f9862b4a926.yaml'
```

The output directory is not created correctly.

File: llm_studio/src/utils/config_utils.py (hunk 4 above)
Comment (maxjeblick) [reply llm_studio/src/utils/config_utils.py]: I added `cfg_dict["problem_type"] = "text_causal_language_modeling"` when saving the yaml.
Dynamic mapping to different problem types should thus be possible. `load_config_yaml` can be generalized once we decide to add additional configurations.

**Review by pascal-pfeiffer — APPROVED** [review]
Thank you for addressing all open issues. lgtm!

**Pull request merged**